bluemira/equilibria/coils/_coil.py

```diff
@@ -0,0 +1,575 @@
+
+    @discretisation.setter
+    def discretisation(self, value: float):
+        """Set coil discretisation"""
+        self._discretisation = np.clip(float(value), EPS, 1)
+        self._discretise()
+
+    def assign_material(
+        self,
+        j_max=NBTI_J_MAX,
+        b_max=NBTI_B_MAX,
+    ) -> None:
+        """
+        Assigns EM material properties to coil
+
+        Parameters
+        ----------
+        j_max: float
+            Overwrite default constant material max current density [A/m^2]
+        b_max: float
+            Overwrite default constant material max field [T]
+
+        Notes
+        -----
+        Will always modify both j_max and b_max of the coil with either the default
+        or specified values.
+
+        """
+        self.j_max = j_max
+        self.b_max = b_max
+
+    def get_max_current(self):
+        """Get max current"""
+        return (
+            np.infty
+            if np.isnan(self.j_max)
+            else get_max_current(self.dx, self.dz, self.j_max)
+        )
+
+    def _discretise(self):
+        """
+        Discretise a coil for greens function magnetic field calculations
+
+        Notes
+        -----
+        Only discretisation method currently implemented is rectangular.
+
+        Possible improvement: multiple discretisations for different coils
+
+        """
+        self._quad_x = np.array([self.x])
+        self._quad_z = np.array([self.z])
+        self._quad_dx = np.array([self.dx])
+        self._quad_dz = np.array([self.dz])
+        self._quad_weighting = np.ones_like(self._quad_x)
+        self._einsum_str = "...j, ...j -> ..."
+
+        if self.discretisation < 1:
+            # How fancy do we want the mesh or just smaller rectangles?
+            self._rectangular_discretisation()
+
+    def _validate_size(self):
+        dx_spec = is_num(self.dx)
+        dz_spec = is_num(self.dz)
+        dxdz_spec = dx_spec and dz_spec
+
+        if (dx_spec ^ dz_spec) and not dxdz_spec:
+            # Check that we don't have dx = None and dz = float or vice versa
+            raise EquilibriaError("Must specify either dx and dz or neither.")
+        if dxdz_spec:
+            # If dx and dz are specified, we presume the coil size should
+            # remain fixed
+            self.fix_size()
+
+            self._set_coil_attributes()
+            self._discretise()
+        else:
+            if not is_num(self.j_max):
+                # Check there is a viable way to size the coil
+                raise EquilibriaError("Must specify either dx and dz or j_max.")
+
+            self._flag_sizefix = False
+
+    def _set_coil_attributes(self):
+        self._current_radius = 0.5 * np.hypot(self.dx, self.dz)
+        self._x_boundary, self._z_boundary = self._make_boundary(
+            self.x, self.z, self.dx, self.dz
+        )
+
+    def _rectangular_discretisation(self):
+        """
+        Discretise a coil into smaller rectangles based on the length
```

Comment:
This is more of a "filament" approach, as each sub-rectangle (as plotted) just has one current filament at its centre. There is probably a mathematical fancy name for something like this. Maybe like "uniform quadrature" except not that.